cpp/include/cudf/io/detail/data_casting.cuh

```diff
@@ -0,0 +1,401 @@
+
+  // Whether in the original JSON this was a string value enclosed in quotes
+  // ({"a":"foo"} vs. {"a":1.23})
+  char const quote_char     = '"';
+  char const backslash_char = '\\';
+
+  // String values are indicated by keeping the quote character
+  bool const is_string_value =
+    num_in_chars >= 2LL && (*in_begin == quote_char) && (*thrust::prev(in_end) == quote_char);
+
+  // Copy literal/numeric value
+  if (not is_string_value) {
+    while (in_begin != in_end) {
+      *out_it++ = *in_begin++;
+    }
+    return {in_begin, out_it, NOT_NULL_FLAG, NO_ERROR_FLAG};
+  }
+
+  // Escape-flag, set after encountering a backslash character
+  bool escape = false;
+
+  // Exclude beginning and ending quote chars from string range
+  if (!options.keepquotes) {
+    ++in_begin;
+    --in_end;
+  }
+
+  // Iterate over the input
+  while (in_begin != in_end) {
+    // Copy single character to output
+    if (!escape) {
+      escape = (*in_begin == backslash_char);
+      if (!escape) { *out_it++ = *in_begin; }
+      in_begin++;
+      continue;
+    }
+
+    // Previous char indicated beginning of escape sequence
+    // Reset escape flag for next loop iteration
+    escape = false;
+
+    // Check the character that is supposed to be escaped
+    auto escaped_char = get_escape_char(*in_begin);
+
+    // We escaped an invalid escape character -> "fail"/null for this item
+    if (escaped_char == NON_ESCAPE_CHAR) { return {in_begin, out_it, NULL_FLAG, INVALID_FLAG}; }
+
+    // Regular, single-character escape
+    if (escaped_char != UNICODE_SEQ) {
+      *out_it++ = escaped_char;
+      ++in_begin;
+      continue;
+    }
+
+    // This is an escape sequence of a unicode code point: \uXXXX,
+    // where each X in XXXX represents a hex digit
+    // Skip over the 'u' char from \uXXXX to the first hex digit
+    ++in_begin;
+
+    // Make sure that there's at least 4 characters left from the
+    // input, which are expected to be hex digits
+    if (thrust::distance(in_begin, in_end) < UNICODE_HEX_DIGIT_COUNT) {
+      return {in_begin, out_it, NULL_FLAG, INVALID_FLAG};
+    }
+
+    auto hex_val = parse_unicode_hex(in_begin);
+
+    // Couldn't parse hex values from the four-character sequence -> "fail"/null for this item
+    if (hex_val < 0) { return {in_begin, out_it, NULL_FLAG, INVALID_FLAG}; }
+
+    // Skip over the four hex digits
+    thrust::advance(in_begin, UNICODE_HEX_DIGIT_COUNT);
+
+    // If this may be a UTF-16 encoded surrogate pair:
+    // we expect another \uXXXX sequence
+    int32_t hex_low_val = 0;
+    if (thrust::distance(in_begin, in_end) >= NUM_UNICODE_ESC_SEQ_CHARS &&
+        *in_begin == backslash_char && *thrust::next(in_begin) == 'u') {
+      // Iterator that skips over '\' and 'u' chars (not yet advancing in_begin, as it may turn out
+      // to not be a surrogate pair)
+      auto low_surrogate_digit_it = thrust::next(thrust::next(in_begin));
```

Comment:
```suggestion
      auto low_surrogate_digit_it = thrust::advance(in_begin, 2);
```

Reply:
You could also use `thrust::next(in_begin, 2)` in place of `thrust::advance` here and above.

Reply:
Yeah, `thrust::next(in_begin, 2)` is much nicer than `thrust::next(thrust::next())` 🙈 😅 
I think we need `next` (rather than `advance`) here, as we don't want to modify the iterator itself; just want to get an iterator that's "advanced" by two.